src/services/events/index.ts

```diff
@@ -0,0 +1,123 @@
+import { CartDto, Product, User } from '../../interfaces'
+
+type EventCallbackFunction<Payload = any> = (payload: Payload) => void
+
+type GetTypeFromInterface<T> = T extends EventCallbackFunction<infer Payload> ? Payload : undefined
+
+type EventsListenerService = () => {
+  emit: <Key extends keyof Events>(
+    event: Key,
+    payload?: GetTypeFromInterface<Events[Key][0]> | undefined,
+  ) => void
+  on: (event: keyof Events, cb: EventCallbackFunction) => void
```

Comment:
Tutaj calback też musi być typowany - aktualnie zawsze jest any a powinno być to samo co jest przekazywane w emit